File: epple2/src/throttle.cpp
/***************************************************************************
 *   Copyright (C) 2008 by Chris Mosher,,,   *
 *   chris@mosher.mine.nu   *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/
#include "throttle.h"
#include "timinggenerator.h"
#include "util.h"


Throttle::Throttle():
	ticksPrev(clock()),
	times(0),
	suspend(false)
{
}

Throttle::~Throttle()
{
}


void Throttle::tick()
{
	suspendIfNecessary();
	throttleIfNecessary();
}
#include <iostream>
void Throttle::throttleIfNecessary()
{
	/*
	 * Check every 100 milliseconds to see how far
	 * ahead we are, and sleep by the difference.
	 */
	++this->times;
	if (this->times >= CHECK_EVERY_CYCLE)
	{
		const clock_t ticksActual = clock()-this->ticksPrev;
		const long ticksDelta = EXPECTED_TICKS-ticksActual;
//		this->speedRatio = (float)ticksDelta/ticksActual;
//		std::cout << speedRatio << std::endl;
		sleep(ticksDelta);

		this->ticksPrev = clock();
		this->times = 0;
	}
}

void Throttle::sleep(const long ticksDelta)
{
	if (ticksDelta > 0)
	{
#ifdef _MSC_VER
		clock_t ticksEnd = clock()+ticksDelta;
		while (clock() < ticksEnd)
		{
			// just spin here
		}
#else
		struct timespec sleepTime;
		sleepTime.tv_sec = ticksDelta/CLOCKS_PER_SEC;
		sleepTime.tv_nsec = ticksDelta*NANOS_PER_CLOCK;
		nanosleep(&this->sleepTime,0);
#endif
	}
}

void Throttle::suspendIfNecessary()
{
/*	while (this->suspend.get())
	{
		try
		{
			this->suspend.wait();
		}
		catch (InterruptedException e)
		{
			Thread.currentThread().interrupt();
		}
	}
*/}

void Throttle::toggleSuspend()
{
	this->suspend = !this->suspend;
//	this->suspend.notifyAll();
}

float Throttle::getSpeedRatio()
{
	return this->speedRatio;
}

File: android/ja2/src/chipset/Throttle.java
/*
 * Created on Jan 4, 2008
 */
package chipset;



import java.util.concurrent.atomic.AtomicBoolean;

import android.util.Log;
import util.Util;



public class Throttle
{
	private static final String LOG_TAG = Throttle.class.getName();
	private static final int CHECK_EVERY_FRACT = 10;
	private static final int CHECK_EVERY_CYCLE = Util.divideRound(TimingGenerator.AVG_CPU_HZ,CHECK_EVERY_FRACT);
	private static final int EXPECTED_MS = 1000/CHECK_EVERY_FRACT;

	private AtomicBoolean suspend = new AtomicBoolean();

	private long msPrev = System.currentTimeMillis();
	private long times;

	private volatile float speedRatio;


	public void tick()
	{
		suspendIfNecessary();
		throttleIfNecessary();
	}

	private void throttleIfNecessary()
	{
		/*
		 * Check every 100 milliseconds to see how far
		 * ahead we are, and sleep by the difference.
		 */
		++this.times;
		if (this.times >= CHECK_EVERY_CYCLE)
		{
			final long msActual = System.currentTimeMillis()-this.msPrev;
			final long msDelta = EXPECTED_MS-msActual;
			this.speedRatio = (float)EXPECTED_MS/msActual;
//			System.err.println(this.speedRatio);

			//			sleep(msDelta);
			// TODO TESTING:
			try {
				Thread.sleep(1);
			} catch (InterruptedException e) {
				Thread.currentThread().interrupt();
			}

			this.msPrev = System.currentTimeMillis();
			this.times = 0;
		}
	}

	private static void sleep(final long msDelta)
	{
		if (msDelta > 0)
		{
			try
			{
				Log.i(LOG_TAG, "Sleeping "+msDelta);
				Thread.sleep(msDelta);
			}
			catch (InterruptedException e)
			{
				Thread.currentThread().interrupt();
			}
		}
	}

	private void suspendIfNecessary()
	{
		// TODO need to wake up if user exits program
		synchronized (this.suspend)
		{
			while (this.suspend.get())
			{
				try
				{
					this.suspend.wait();
				}
				catch (InterruptedException e)
				{
					Thread.currentThread().interrupt();
				}
			}
		}
	}

	public void toggleSuspend()
	{
		synchronized (this.suspend)
		{
			this.suspend.set(!this.suspend.get());
			this.suspend.notifyAll();
		}
	}

	public float getSpeedRatio()
	{
		return this.speedRatio;
	}
}
